### PR TITLE
User data feature

### PR DESCRIPTION
**Pull request by B-Reif**
This PR adds some support for parsing user data chunks and attaching them to other data. 

Changes include:
- Adds optional user data field to Cel, Layer, Tag, and AsepriteFile structs.
- Adds a new `UserDataContext` enum to `ParseInfo`, which stores some information about the most-recently parsed chunk. When parsing a UserData chunk, the context is used to attach the user data to other existing content.
- Adds a new `InternalError` case to the `AsepriteParseError` enum. This would only be returned in case of some violated invariant. For now, it's only returned if the `UserDataContext` holds a bad id.
- Refactors `ParseInfo` to store layers during parsing, to support attaching user data.
- Refactors some methods to use `CelId` more.

A few questions:
- The Aseprite file spec mentions that a [user data chunk can follow any other chunk](https://github.com/aseprite/aseprite/blob/master/docs/ase-file-specs.md#user-data-chunk-0x2020). The Aseprite UI only exposes user data fields in some of these contexts. What kinds of chunks should we support user data on?
- Currently the 'color' field in `UserData` is an array of bytes. Since we're already exposing `image::Rgba`, should we change this field to that type?

### CODE REVIEW

File: src/parse.rs [hunk 1]
```diff
@@ -332,7 +490,31 @@ impl Chunk {
...
+        mut bytes_available: i64,
+        reader: &mut AseReader<R>,
+    ) -> Result<Vec<Self>> {
+        let mut chunks: Vec<Chunk> = Vec::new();
+        for _idx in 0..count {
+            let chunk = Self::read(&mut bytes_available, reader)?;
+            // Attach any user data chunks to the previously read chunk. Otherwise, push chunk to Vec and continue.
+            chunks.push(chunk);
+            // if let ChunkType::UserData = chunk.chunk_type {
+            //     let user_data = user_data::parse_userdata_chunk(&chunk.content.data)?;
```
Comment (alpine-alpaca) [inline src/parse.rs:506]: I'm guessing that's not needed anymore? And the comment above may also not be relevant anymore?

File: src/parse.rs [hunk 2]
```diff
@@ -1,40 +1,187 @@
...
 use crate::tileset::{Tileset, TilesetsById};
+use crate::user_data::UserData;
 use crate::{error::AsepriteParseError, AsepriteFile, PixelFormat};
 use log::debug;
 use std::io::Read;
 
 use crate::Result;
 use crate::{cel, color_profile, layer, palette, slice, tags, user_data, Tag};
 
+enum LayerParseInfo {
```
Comment (alpine-alpaca) [inline src/parse.rs:15]: Can you maybe add a comment what the idea is behind this type?

**Comment by alpine-alpaca** [comment]
Regarding your questions:

- We only need to support the user data chunks that are actually exposed by the UI. I can imagine users storing some meta data for custom tools in these chunks, but I expect files to only be authored by hand. So, no need to support everything that the file format theoretically allows. If someone has a specific use case that's currently not supported they'll create an issue and we go from there.

- The color in user data should be an `Rgba` since we're already exposing that through other parts of the API.

File: src/parse.rs (hunk 1 above)
Comment (B-Reif) [reply src/parse.rs:506]: Oh yes, good catch. Previously I did this a little differently (per the above comment) but moved to the chunk context field instead.

File: src/parse.rs (hunk 2 above)
Comment (B-Reif) [reply src/parse.rs:15]: Prior to this change, the layer data in the first frame was stored in the `found_layers` vec. After the first frame was done, that data moved into LayersData`, and then into `ParseInfo. I moved `found_layers` into the ParseData so we can attach UserData to it without accessing a separate variable.

I suppose we could actually remove LayersData from ParseInfo, put `layer_parents: Option<_>` on the ParseInfo directly, and set it to `Some(parents)`. For now I've left a comment explaining how it works.

**Comment by B-Reif** [comment]
Ok, made some changes:
- removed commented-out old code
- use image::Rgba in UserData
- Removed the LayerParseInfo struct. The `Vec<LayerData>` is now converted to `LayersData` during validation.
- Added tests for UserData colors

**Comment by alpine-alpaca** [comment]
Can I close this since you created the merged PR?

**Pull request merged**